### PR TITLE
spack templates BIS: all configs available in spack.yaml and its dir

### PR DESCRIPTION
**Pull request by marcodelapierre**
Hi @pditommaso, this PR updates the Spack templates for Docker/Singularity, so that all package build configurations are visible in `spack.yaml` and its sister yaml, `compilers.yaml`, in the `/opt/spack` directory. 

This change does not impact the way packages are compiled, but in my opinion is beneficial to users as it improved transparency of how the packages are built, hence improving provenance of the build from the software perspective.

@bebosudo FYI

### CODE REVIEW

**Comment by marcodelapierre** [comment]
CI Build fails to be investigated yet (will tackle fails in https://github.com/seqeralabs/wave/pull/369 first)

File: src/main/resources/io/seqera/wave/spack/spack-builder-dockerfile.txt [hunk 1]
```diff
@@ -16,32 +16,33 @@ ENV \
...
-         -e 's;^ *flags: *{};    flags:\n      cflags: -O3\n      cxxflags: -O3\n      fflags: -O3;' \
-         /root/.spack/linux/compilers.yaml \
-&& cd /opt/spack-env \
-&& spack config add config:install_tree:/opt/software \
-&& spack config add concretizer:unify:true \
-&& spack config add concretizer:reuse:true \
-&& spack config add packages:all:target:[{{spack_arch}}] \
-&& printf "  view: /opt/view\n" >> /opt/spack-env/spack.yaml
+# Set up Spack environment
+RUN cd /opt/spack-env \
```
Comment (bebosudo) [inline src/main/resources/io/seqera/wave/spack/spack-builder-dockerfile.txt:20]: The `/opt/spack-env` directory doesn't seem to exist in the current `spack/ubuntu-jammy:develop` image, why are we removing the `mkdir`?

```
[alberto@fedora]$ podman run --rm -it --entrypoint bash spack/ubuntu-jammy:develop 
root@b26a32e6d4e9:~# 
root@b26a32e6d4e9:~# cd /opt/spack-env
bash: cd: /opt/spack-env: No such file or directory
```

File: src/main/resources/io/seqera/wave/spack/spack-builder-singularityfile.txt [hunk 2]
```diff
@@ -20,27 +20,26 @@ EOF
...
     . /aws-env.sh
 
-    # Copy and modify spack.yaml
-    sed -i -e 's;compilers:;compilers::;' \
-           -e 's;^ *flags: *{};    flags:\n      cflags: -O3\n      cxxflags: -O3\n      fflags: -O3;' \
-           /root/.spack/linux/compilers.yaml
-
     # Set up Spack environment
     export PATH=/opt/spack/bin:$PATH
     cd /opt/spack-env
```
Comment (bebosudo) [inline src/main/resources/io/seqera/wave/spack/spack-builder-singularityfile.txt:25]: How was this `cd` working at all if the directory doesn't exist in the image?

**Review by bebosudo — APPROVED** [review]

File: src/main/resources/io/seqera/wave/spack/spack-builder-dockerfile.txt (hunk 1 above)
Comment (marcodelapierre) [reply src/main/resources/io/seqera/wave/spack/spack-builder-dockerfile.txt:20]: The directory gets created at line 3 of the Dockerfile by this `COPY` command:
```
COPY spack.yaml /opt/spack-env/spack.yaml
```

File: src/main/resources/io/seqera/wave/spack/spack-builder-singularityfile.txt (hunk 2 above)
Comment (marcodelapierre) [reply src/main/resources/io/seqera/wave/spack/spack-builder-singularityfile.txt:25]: same reason as above, The directory gets created at line 3 of the Dockerfile by this `COPY` command:
```
COPY spack.yaml /opt/spack-env/spack.yaml
```

**Comment by marcodelapierre** [comment]
CI fixed.

Ready for review @pditommaso

**Comment by marcodelapierre** [comment]
@pditommaso  this is now fully ready for review (after CI fix)

**Comment by pditommaso** [comment]
To make this work both for Docker and Singularity it has been a true bloodbath. How this has been tested?

**Comment by marcodelapierre** [comment]
Sister PR https://github.com/seqeralabs/wave/pull/371 locally merged into this one

**Comment by marcodelapierre** [comment]
@bebosudo  @pditommaso  Testing update:  

I have tested the  Dockerfile and Singularity def file as generated by Wave with this PR. I have extracted the Dockerfile/Singularityfile and used them to build a simple package, `cowsay`, locally, using Docker and Singularity. 

The test was successful:
- final spack.yaml included all the changes
- crucially, the builds were successful

I do not expect any further updates on these files, excepty those that may be required for bug fixing (the current production Wave does NOT produce working Spack containers).

@pditommaso I suggest any additional review by you, then merge.

**Review by pditommaso — APPROVED** [review]

**Comment by pditommaso** [comment]
@marcodelapierre conflicts to be solved 👆

**Pull request merged**